runtime/queries/column_time_grain.go

```diff
@@ -0,0 +1,128 @@
+package queries
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+)
+
+type ColumnTimeGrain struct {
+	TableName  string
+	ColumnName string
+	Result     runtimev1.TimeGrain
+}
+
+var _ runtime.Query = &ColumnTimeGrain{}
+
+func (q *ColumnTimeGrain) Key() string {
+	return fmt.Sprintf("ColumnTimeGrain:%s:%s", q.TableName, q.ColumnName)
+}
+
+func (q *ColumnTimeGrain) Deps() []string {
+	return []string{q.TableName}
+}
+
+func (q *ColumnTimeGrain) MarshalResult() any {
+	return q.Result
+}
+
+func (q *ColumnTimeGrain) UnmarshalResult(v any) error {
+	res, ok := v.(runtimev1.TimeGrain)
+	if !ok {
+		return fmt.Errorf("ColumnTimeGrain: mismatched unmarshal input")
+	}
+	q.Result = res
+	return nil
+}
+func (q *ColumnTimeGrain) Resolve(ctx context.Context, rt *runtime.Runtime, instanceID string, priority int) error {
+	sampleSize := int64(500000)
+	cq := &TableCardinality{
+		TableName: q.TableName,
+	}
+	err := rt.Query(ctx, instanceID, cq, priority)
+	if err != nil {
+		return err
+	}
+	var useSample string
+	if sampleSize > cq.Result {
+		useSample = ""
+	} else {
+		useSample = fmt.Sprintf("USING SAMPLE %d ROWS", sampleSize)
+	}
```

Comment:
Probably it was slight optimisation because as the query plan says the db still creates some sampling step:
```
D create table rng as (select * from range(0,10));
D explain select * from rng using sample 11 rows;

┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│      RESERVOIR_SAMPLE     │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          11 rows          │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            rng            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           range           │
│           EC=10           │
└───────────────────────────┘                             
D explain select * from rng using sample 20 rows;

┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│      RESERVOIR_SAMPLE     │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          20 rows          │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            rng            │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           range           │
│           EC=10           │
└───────────────────────────┘                    
```

Reply:
Got it, then let's keep the conditional sampling clause to be safe